src/Interpreters/Context.cpp

```diff
@@ -3810,6 +3811,43 @@ void Context::clearTextIndexDictionaryBlockCache() const
         shared->text_index_dictionary_block_cache->clear();
 }
 
+void Context::setTextIndexHeaderCache(const String & cache_policy, size_t max_size_in_bytes, size_t max_entries, double size_ratio)
+{
+    std::lock_guard lock(shared->mutex);
+
+    if (shared->text_index_header_cache)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Text index header cache has been already created.");
+
+    shared->text_index_header_cache = std::make_shared<TextIndexHeaderCache>(cache_policy, max_size_in_bytes, max_entries, size_ratio);
+}
+
+void Context::updateTextIndexHeaderCacheConfiguration(const Poco::Util::AbstractConfiguration & config)
+{
+    std::lock_guard lock(shared->mutex);
+
+    if (!shared->text_index_header_cache)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Text index header cache was not created yet.");
+
+    size_t max_size_in_bytes = config.getUInt64("text_index_header_cache_size", DEFAULT_TEXT_INDEX_HEADER_CACHE_MAX_SIZE);
+    size_t max_entries = config.getUInt64("text_index_header_cache_max_entries", DEFAULT_TEXT_INDEX_HEADER_CACHE_MAX_ENTRIES);
+    shared->text_index_header_cache->setMaxSizeInBytes(max_size_in_bytes);
+    shared->text_index_header_cache->setMaxCount(max_entries);
+}
+
+std::shared_ptr<TextIndexHeaderCache> Context::getTextIndexHeaderCache() const
+{
+    SharedLockGuard lock(shared->mutex);
+    return shared->text_index_header_cache;
+}
+
+void Context::clearTextIndexHeaderCache() const
+{
+    auto cache = getTextIndexHeaderCache();
+
+    /// Clear the cache without holding context mutex to avoid blocking context for a long time
+    if (cache)
+        cache->clear();
```

Comment:
You could be calling clear and `getTextIndexHeaderCache` at the same time? Or is there some other protection?